apollo-federation/src/query_plan/operation.rs

```diff
@@ -2222,8 +2283,179 @@ impl SelectionSet {
+        let Some((first, rest)) = updates.split_first() else {
+            // PORT_NOTE: The TypeScript version asserts here.
+            return Err(FederationError::internal(
+                "Should not be called without any updates",
+            ));
+        };
+        if rest.is_empty() {
+            // Optimize for the simple case of a single selection, as we don't have to do anything
+            // complex to merge the sub-selections.
+            return first.rebase_on(
+                parent_type,
+                /*named_fragments*/ &Default::default(),
+                schema,
+                RebaseErrorHandlingOption::ThrowError,
+            );
+        }
+
+        let element = first.element()?.rebase_on(
+            parent_type,
+            schema,
+            RebaseErrorHandlingOption::ThrowError,
+        )?;
+        let Some(element) = element else {
+            return Err(FederationError::internal(
+                "Unable to rebase selection updates",
+            ));
+        };
+        let sub_selection_parent_type: Option<CompositeTypeDefinitionPosition> = match element {
+            OpPathElement::Field(ref field) => field.data().base_type()?.try_into().ok(),
+            OpPathElement::InlineFragment(ref inline) => Some(inline.data().casted_type()),
+        };
+
+        let Some(ref sub_selection_parent_type) = sub_selection_parent_type else {
+            // This is a leaf, so all updates should correspond ot the same field and we just use the first.
+            return Ok(Some(selection_of_element(
+                element, /*sub_selection*/ None,
+            )?));
+        };
+
+        // This case has a sub-selection. Merge all sub-selection updates.
+        let mut sub_selection_updates: MultiMap<SelectionKey, Selection> = MultiMap::new();
+        for update in updates {
+            if let Some(sub_selection_set) = update.selection_set()? {
+                sub_selection_updates.extend(
+                    sub_selection_set
+                        .selections
+                        .iter()
+                        .map(|(k, v)| (k.clone(), v.clone())),
+                );
+            }
+        }
+        let updated_sub_selection = Some(Self::make_selection_set(
+            schema,
+            sub_selection_parent_type,
+            sub_selection_updates,
+        )?);
+        Ok(Some(selection_of_element(element, updated_sub_selection)?))
+    }
+
+    fn make_selection_set(
+        schema: &ValidFederationSchema,
+        parent_type: &CompositeTypeDefinitionPosition,
+        updated_selections: MultiMap<SelectionKey, Selection>,
+    ) -> Result<SelectionSet, FederationError> {
+        let mut selections = SelectionMap::new();
+        for (_key, updates) in updated_selections.into_iter() {
+            if let Some(selection) = Self::make_selection(schema, parent_type, updates)? {
+                selections.insert(selection);
+            }
+        }
```

Comment:
I need this method to take a iterator over slices for my `add_at_path` work. This will require some unfortunate changes to the iterators that the calls (I opened an issue on the multimap repo for better iterators, but th
```suggestion
    fn make_selection_set<'a, I>(
        schema: &ValidFederationSchema,
        parent_type: &CompositeTypeDefinitionPosition,
        updated_selections: I, 
    ) -> Result<SelectionSet, FederationError>
    where
        I: IntoIterator<Item = &'a [Selection]>,
    {
        let mut selections = SelectionMap::new();
        for updates in updated_selections.into_iter() {
            if let Some(selection) = Self::make_selection(schema, parent_type, updates)? {
                selections.insert(selection);
            }
        }
```